Add support for Arduino UNO R4

soburi opened the pull request:
Adds support for Arduino UNO R4.

I have verified that hello_world, blinky, and button sample work with this PR.

We can flash and debug with the cmsis-dap adapter and pyOCD.
But not supporting flash program with dfu-util, because Arduino IDE use not standard version of dfu-util.

To avoid complicating initial code for supporting the SoC,
I have implemented only the bare minimum for now.

~~We can't use [FSP](https://github.com/renesas/fsp) that the BSP Renesas provides is not an OSS license (It doesn't allow execution outside of Renesas SoCs).~~
~~I write this PR entirely from scratch, so it is not dependent on FSP.~~
~~But it is good to use FSP for reliability reasons if we can use it.~~

~~This PR is still remain some TODOs.~~
~~Because it still needs to complete implementing some functions (clock_control, pinctrl, etc.), mark them as TODO,~~
~~depending on it.~~

~~This PR changes R-Car UART driver to support RA UART.~~
~~But I don't have R-Car board, so I can't test it.~~

~~If Renesas plans to support Zephyr officially, I would prefer to follow that policy. (But that doesn't mean we can't proceed without it.)~~

soburi committed:
drivers: gpio: Add GPIO driver for Renesas RA series

Add initial support for Renesas RA GPIO.

To avoid complicating the initial code for supporting the SoC,
I have implemented only the bare minimum for now.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
drivers: serial: Add UART driver for Renesas RA series

Adding initial support for Renesas RA UART.

To avoid complicating initial code for supporting the SoC,
I have implemented only the bare minimum for now.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
boards: arm: add initial support for Arduino UNO R4 Minima

Add support for the Arduino UNO R4 Minima.
This board is the newest version of Arduino that uses Renesas RA4M1 SoC.

This commit provides only limited support to simplify the initial
support patch.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
soc: arm: add support for Renesas RA4M1 series SoC

Add essential support for RA4M1 Series.
It only defines `r7fa4m1ab3cfm` currently.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
drivers: pinctrl: Add pinctrl driver for Renesas RA series

To avoid complicating the initial code for supporting the SoC,
I have implemented only the bare minimum for now.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
drivers: clock_control: Add clock driver for Renesas RA series

Add initial support for Renesas RA clock generation circuit.

It returns a fixed value to simplify the first commit to get the UART
working now.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

fabiobaltieri commented:
> If Renesas officially supports Zephyr, I would prefer to follow that policy. However if not so, proceeding on a community basis is necessary. Has anyone heard the intentions of Renesas?

There's a HAL already for SmartBond devices (https://github.com/zephyrproject-rtos/hal_renesas) and code for both that and R-Car.

@andrzej-kaczmarek @blauret @aaillet @lorc any idea if there's interest to have official support for RA family?

lorc commented:
My team is mostly interested in CortexA cores and virtualization, so I'm afraid no one from my side can help with this.

blauret commented:
It's great initiative. Just a few comments:

- Debugging is supported with JLink (See [R7FA4M1AB](https://www.segger.com/supported-devices/renesas/ra4m1)) . You don't need the Renesas debug server. JLink is widely accepted debug option in the Zephyr environment. 
- For the licensing side of the FSP package, I'm not sure how it works but Platformio does provide and [RA package](https://github.com/platformio/platform-renesas-ra). If think they figured a way to support that environment, it should be possible to comply with Zephyr community requirements. 

My team is focused on the Smartbond devices. I'll try to reach out in the organization to see if someone can help. That said, it's en of July and it's summer holidays it could take some time.

soburi commented:
@fabiobaltieri @lorc @blauret 

Thank you for responses.

> - Debugging is supported with JLink (See [R7FA4M1AB](https://www.segger.com/supported-devices/renesas/ra4m1)) . You don't need the Renesas debug server. JLink is widely accepted debug option in the Zephyr environment.

I misunderstood the need for Renesas-specific software for debugging.
I verify it can debug with JLink and pyOCD.
I made pyocd the default because cmsis-dap adapters are easy to obtain.

> - For the licensing side of the FSP package, I'm not sure how it works but Platformio does provide and [RA package](https://github.com/platformio/platform-renesas-ra). If think they figured a way to support that environment, it should be possible to comply with Zephyr community requirements.

At a glance, [RA package](https://github.com/platformio/platform-renesas-ra) contains no FSP sources. FSP sources are in the separated repository [framework-renesas-fsp](https://registry.platformio.org/tools/platformio/framework-renesas-fsp).
So, It still needs consideration...


> My team is mostly interested in CortexA cores and virtualization, so I'm afraid no one from my side can help with this.

> My team is focused on the Smartbond devices. I'll try to reach out in the organization to see if someone can help.

Thanks a lot.
I know there are some things difficult in large organization.
Could you tell me any news (or rumor?) if you get in future.

> That said, it's en of July and it's summer holidays it could take some time.

Have a nice holidays!

aaillet commented:
Hi,

I'm afraid that as people from other teams, I'm mainly focusing on my part, Cortex-R support for R-Car SoCs.

From my point of view, Smartbond code and HAL will not fit RA series as Smartbond is coming from a brand that Renesas bought. As this series was already existing before it's brand acquirement, I think it's not using Renesas IPs.

On the other hand, it may be possible to reuse R-CAR code if same IPs are used.

Concerning interest from Renesas to officially support Zephyr on RA/RZ this series, I'll ask my internal contact if plans are existing !

gmarull commented:
> > If Renesas officially supports Zephyr, I would prefer to follow that policy. However if not so, proceeding on a community basis is necessary. Has anyone heard the intentions of Renesas?
> 
> There's a HAL already for SmartBond devices (https://github.com/zephyrproject-rtos/hal_renesas) and code for both that and R-Car.

It would be nice to keep implementation native, HALs are not mandatory and often more a pain than anything else.

aaillet commented:
> This PR is still remain some TODOs. Because it still needs to complete implementing some functions (clock_control, pinctrl, etc.), mark them as TODO, depending on it.
> 
> This PR changes R-Car UART driver to support RA UART. But I don't have R-Car board, so I can't test it.
> 

As there is still some work to do (I'm especially talking about that we need to find smart way to support both ra and rcar in serial driver) and test to do on board, is it ok for you to convert this PR to draft ?

soburi commented:
@aaillet 

> As there is still some work to do (I'm especially talking about that we need to find smart way to support both ra and rcar in serial driver) and test to do on board, is it ok for you to convert this PR to draft ?

I got it. I reverted it to draft.

gmarull commented on drivers/clock_control/clock_control_ra.c:
```diff
@@ -0,0 +1,249 @@
+	volatile uint32_t MSTPCRE;
+};
+
+struct system_regs {
+	volatile uint8_t UNUSED0[32];
+	volatile uint32_t SCKDIVCR;
+	volatile uint8_t UNUSED1[2];
+	volatile uint8_t SCKSCR;
+	volatile uint8_t UNUSED2;
+	volatile uint16_t PLLCCR;
+	volatile uint8_t PLLCR;
+	volatile uint8_t PLLCCR2;
+	volatile uint8_t UNUSED3[5];
+	volatile uint8_t MEMWAIT;
+	volatile uint8_t MOSCCR;
+	volatile uint8_t UNUSED4[3];
+	volatile uint8_t HOCOCR;
+	volatile uint8_t UNUSED5[5];
+	volatile uint8_t OSCSF;
+	volatile uint8_t UNUSED6[99];
+	volatile uint8_t OPCCR;
+	volatile uint8_t UNUSED7[4];
+	volatile uint8_t HOCOWTCR;
+	volatile uint8_t UNUSED8[856];
+	volatile uint16_t PRCR;
+	volatile uint8_t UNUSED9[128];
+	volatile uint8_t SOSCCR;
+	volatile uint8_t UNUSED10[639];
+};
```
suggestion: to improve code readability and consistency with native drivers, better define offsets and use the `sys_write/read` APIs.

soburi replied:
Fixed it.

soburi committed:
drivers: interrupt_controller: Add icu driver for Renesas RA series

To avoid complicating the initial code for supporting the SoC,
I have implemented only the bare minimum for now.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi committed:
tests: arch: arm: arm_irq_vector_table: exclude arduino_uno_r4_minima

Renesas RA always uses interrupt handlers dynamically.
But this test requires static vector tables.
So, it needs to exclude platforms that use Renesas RA.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>

soburi commented:
@aaillet 

Although there is some commonality in design, there are many differences such as register length, so we have given up on sharing scif with rcar at this time.

aaillet commented:
> @aaillet
> 
> Although there is some commonality in design, there are many differences such as register length, so we have given up on sharing scif with rcar at this time.

Yes, it looks complicated to do it without making a mess in the driver atm !
Thanks for trying anyway :+1:

soburi commented on drivers/pinctrl/pinctrl_ra.c:
```diff
@@ -0,0 +1,83 @@
+#define DT_DRV_COMPAT renesas_ra_pinctrl
+
+static const uint32_t PORT_NUM = 15;
+static const uint32_t PIN_NUM = 16;
```
Changed to define as a macro.

soburi commented on drivers/pinctrl/pinctrl_ra.c:
```diff
@@ -0,0 +1,83 @@
+	PWPR_B0WI_POS = 7,
+};
+
+static inline uint32_t PmnFPS_read(size_t port, size_t pin)
```
Changed function names for register read/write to `pinctrl_ra_(read|write)_...`.

soburi commented on drivers/pinctrl/pinctrl_ra.c:
```diff
@@ -0,0 +1,83 @@
+	PmnFPS_write(pinc->port, pinc->pin, pincfg.config);
+}
+
+int ra_pinctrl_query_config(uint32_t port, uint32_t pin, struct ra_pinctrl_soc_pin *const pincfg)
```
I fixed as so.

soburi commented on drivers/interrupt_controller/intc_ra_icu.c:
```diff
@@ -0,0 +1,115 @@
+	IELSR_IR_POS = 16,
+};
+
+const uint8_t IRQCR_IRQMD_MASK = 0x3;
```
I fixed it as such.

soburi commented on drivers/interrupt_controller/intc_ra_icu.c:
```diff
@@ -0,0 +1,115 @@
+#include <zephyr/sw_isr_table.h>
+#include <errno.h>
+
+#define REG DT_INST_REG_ADDR(0)
```
Define `IELSRn_REG(n)` and `IRQCRi_REG(i)`  and cleanup related codes.

soburi commented on drivers/interrupt_controller/intc_ra_icu.c:
```diff
@@ -0,0 +1,115 @@
+			       const void *parameter, uint32_t flags)
+{
+	uint32_t event = ((flags & 0x0000FF00) >> 8);
+	uint32_t intcfg = ((flags & 0x00FF0000) >> 16);
```
Define `RA_ICU_FLAG_EVENT_(OFFSET|MASK)` and `RA_ICU_FLAG_INTCFG_(OFFSET|MASK)`

soburi commented on drivers/gpio/gpio_ra.c:
```diff
@@ -0,0 +1,417 @@
+	pincfg.port = config->port;
+
+	if (flags & GPIO_INT_ENABLE) {
+		uint32_t intcfg = 0;
```
Remove initializer.

soburi commented on drivers/gpio/gpio_ra.c:
```diff
@@ -0,0 +1,417 @@
+	uint16_t setbits = value & mask;
+	uint16_t clrbits = (~value) & mask;
+
+	reg_write(dev, PCNTR3_OFFSET, (clrbits << 16) | setbits);
```
Define `PCNTR3_PORR0_OFFSET` as enum.

soburi commented on drivers/gpio/gpio_ra.c:
```diff
@@ -0,0 +1,417 @@
+{
+	struct gpio_ra_data *data = dev->data;
+	int irq =
+		ra_icu_query_exists_irq(gpio_ra_irq_info_event(data->port_irq_info[port_irq].info));
```
I fixed it as such.

soburi commented on drivers/gpio/gpio_ra.c:
```diff
@@ -0,0 +1,417 @@
+	const struct gpio_ra_config *config = dev->config;
+
+	for (int i = 0; i < config->irq_info_size; i++) {
+		for (int j = 0; j < config->irq_info[i].num; j++) {
```
I fixed it as such.

soburi commented on drivers/serial/Kconfig.ra:
```diff
@@ -0,0 +1,13 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config UART_RA
+	bool "Renesas R-Car UART Driver"
```
Oops!
I changed it to `RA series`.

fabiobaltieri reviewed (commented):
good stuff, few more comments

soburi commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -26,6 +26,15 @@
+		icu: interrupt-controller@40006000 {
+			compatible = "renesas,ra-interrupt-controller-unit";
+			reg = <0x40006000 0x40>;
+			reg-names="icu";
```
Fixed it.

soburi commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -147,6 +147,13 @@
+		pinctrl: pinctrl@40040800 {
+			compatible = "renesas,ra-pinctrl";
+			reg = <0x40040800 0x500 0x40040d03 0x1>;
+			reg-names="pfs","pmisc_pwpr";
```
Fixed it.

soburi commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -26,6 +26,109 @@
+		cgc: cgc@4001e000 {
+			compatible = "renesas,ra-clock-generation-circuit";
+			reg = <0x4001e000 0x40 0x40047000 0x10>;
+			reg-names="system","mstp";
```
Fixed it.

soburi commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -26,6 +26,109 @@
+				#clock-cells = <0>;
+			};
+
+			fclk: fclk{
```
Fixed it.

soburi commented on drivers/gpio/gpio_ra.c:
```diff
@@ -0,0 +1,440 @@
+		DT_INST_FOREACH_PROP_ELEM(idx, interrupt_names, GPIO_RA_IRQ_INFO)};                \
+	static struct gpio_ra_config gpio_ra_config_##idx = {                                      \
+		.common =                                                                          \
+			{                                                                          \
```
Fixed it.

soburi commented on drivers/clock_control/clock_control_ra.c:
```diff
@@ -0,0 +1,284 @@
+	case RA_CLOCK_SCI(0):
+		*rate = pclka_freq;
+		break;
+	}
```
Added default section for returning -EINVAL.

soburi commented on drivers/clock_control/clock_control_ra.c:
```diff
@@ -0,0 +1,284 @@
+
+static void crude_busy_loop(uint32_t usec_to_wait)
+{
+	const uint32_t loops_per_sec = (sys_clock_hw_cycles_per_sec() / 5);
```
It is CPU cycles per busy loop. I rewrote the busy loop by the inline assembler and renamed the define to the `cycles_per_loop` constant.

fabiobaltieri replied:
fancy!

soburi commented on drivers/clock_control/clock_control_ra.c:
```diff
@@ -0,0 +1,284 @@
+	SYSTEM_write8(MOSCCR_OFFSET, !IS_CLOCK_ENABLED(mosc));
+	SYSTEM_write8(HOCOCR_OFFSET, !IS_CLOCK_ENABLED(hoco));
+
+#if IS_CLOCK_ENABLED(sosc)
```
It is a preferable way. I rewrote it with an if **statement**.

carlescufi commented:
@fabiobaltieri could you take a look?

fabiobaltieri commented on drivers/serial/uart_ra.c:
```diff
@@ -0,0 +1,396 @@
+			      PRE_KERNEL_1, CONFIG_SERIAL_INIT_PRIORITY,	\
+			      &uart_ra_driver_api);				\
+										\
+	UART_RCAR_INIT_CFG(n);
```
maybe move this before the `DEVICE_DT_INST_DEFINE`? Then you can drop the forward declaration.

fabiobaltieri commented on boards/arm/arduino_uno_r4/board.cmake:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2023 TOKITA Hiroshi <tokita.hiroshi@fujitsu.com>
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_Args(pyocd "--target=r7fa4m1ab")
```
`board_runner_args` (all lowercase)

soburi replied:
Fixed it.

fabiobaltieri commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -26,6 +26,109 @@
+				compatible = "fixed-factor-clock";
+				status = "disabled";
+				clocks = <&sysclk>;
+				#clock-cells = <0>;
+			};
+
+			pclka: pclka {
+				compatible = "fixed-factor-clock";
+				status = "disabled";
+				clocks = <&sysclk>;
+				#clock-cells = <0>;
+			};
+
+			pclkb: pclkb {
+				compatible = "fixed-factor-clock";
+				status = "disabled";
+				clocks = <&sysclk>;
+				#clock-cells = <0>;
+			};
+
+			pclkc: pclkc {
+				compatible = "fixed-factor-clock";
+				status = "disabled";
+				clocks = <&sysclk>;
+				#clock-cells = <0>;
+			};
+
+			pclkd: pclkd {
+				compatible = "fixed-factor-clock";
+				status = "disabled";
+				clocks = <&sysclk>;
+				#clock-cells = <0>;
+			};
+
+			fclk: fclk {
+				compatible = "fixed-factor-clock";
+				status = "disabled";
+				clocks = <&sysclk>;
+				#clock-cells = <0>;
+			};
```
Wow, this block is super verbose, especially on the board file, you'll end up with a ton of nodes. How about dropping the child noes and just using properties for these? That's what the stm32 rcc block does https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/arduino_giga_r1/arduino_giga_r1.dtsi#L33, you could make them optional and use that instead of the disabled nodes, I think the configuration would become a lot easier to understand.

soburi replied:
The clock source was moved to the `clocks { ...}` section, and the division setting was replaced with the clock controller property.
It has cleared up a lot.

fabiobaltieri commented:
@soburi great job putting this together!

gmarull commented on dts/bindings/pinctrl/renesas,ra-pinctrl.yaml:
```diff
@@ -0,0 +1,32 @@
+
+    include:
+      - name: pincfg-node.yaml
+        property-allowlist:
+          - bias-disable
+          - bias-pull-up
+          - drive-open-drain
+          - input-enable
+          - drive-strength
```
where are these handled? I can't see them being referenced in pinctrl_soc.h

soburi replied:
Not yet implemented these, but the reference manual says, these functions exist.
I am adding these for now as I will implement them in the near future.
Is it better not to enter it at this time?
![ra4m1_pinconf](https://github.com/zephyrproject-rtos/zephyr/assets/458281/6053400d-8984-4248-a4b3-90b3fe057b8a)

gmarull replied:
If they're not used, they shouldn't be allowed yet, this will create confusion to the users.

soburi replied:
Ah, I got it.
I removed it.

gmarull commented on dts/bindings/pinctrl/renesas,ra-pinctrl.yaml:
```diff
@@ -0,0 +1,26 @@
+    Definitions for a pinctrl state.
+  child-binding:
+
+    include:
+      - name: pincfg-node.yaml
```
this should be removed entirely, none of the props are used right?

soburi replied:
I had a little misunderstanding. I deleted it additionally.
Thank you for pointing it out.

gmarull commented on drivers/interrupt_controller/intc_ra_icu.c:
```diff
@@ -0,0 +1,115 @@
+	return irqn;
+}
+
+static int ra_icu_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return 0;
+}
+
+DEVICE_DT_INST_DEFINE(0, ra_icu_init, NULL, NULL, NULL, PRE_KERNEL_1, CONFIG_INTC_INIT_PRIORITY,
```
just pass NULL to the init function, it;s now optional

soburi replied:
I fixed as so.

gmarull commented on dts/arm/renesas/ra/ra-cm4-common.dtsi:
```diff
@@ -115,6 +116,135 @@
+			reg = <0x40040020 0x20>;
+			gpio-controller;
+			#gpio-cells = <2>;
+			port-no = <1>;
```
why is this property needed? this is not hw description?

soburi replied:
Yes, it is an index of gpio.
The index can be calculated from the address.
I replaced it with the calculated value.

fabiobaltieri replied:
fancy, you sure the new one is a good idea? it relies on the port memory space being contiguous (question for both of you @soburi @gmarull)

soburi replied:
In the RA series, all GPIO ports use continuous addresses. Even if there is a dropout in the middle, the port number will not change, so calculation is possible.

CM4-based SoC uses 0x40040000 to 0x40040160 for ioport0 .. ioportb.
In the case of CM33-based SoC, it also uses 0x40080000 to 0x40080160.

gmarull replied:
In past situations, we've always restricted DT to the hardware description itself (port-no is needed here because of how software works, not because it describes hw uniquely).

gmarull commented on drivers/gpio/gpio_ra.c:
```diff
@@ -0,0 +1,439 @@
+	return gpio_ra_pin_configure(dev, pin, pincfg | mode | trig);
+}
+
+static int gpio_ra_init(const struct device *dev)
+{
+	return 0;
+}
```
ditto

soburi replied:
Fixed it.

eGit commented:
What's keeping Zephyr from supporting Arduino Uno R3? Not enough RAM?

soburi commented:
@eGit 

> What's keeping Zephyr from supporting Arduino Uno R3? Not enough RAM?

ZephyrRTOS intends 32bit/64bit MPUs. I think it may difficult to port 16bit/8bit.
